### ISSUE TITLE
Module not work

### ISSUE THREAD

**Issue by BaronMsk** [puppetlabs/puppetlabs-kubernetes#585]
## Describe the Bug
After commit: e7addfbeaf3ff75496e31c2f44f22355f75dc6a3 not work modules
log join node
Notice: /Stage[main]/Main/Exec[kubeadm join]/returns: unknown flag: --config '/etc/confi.yaml' --ignore-preflight-errors 'Service-Docker'
Notice: /Stage[main]/Main/Exec[kubeadm join]/returns: To see the stack trace of this error execute with --v=5 or higher
Error: '["kubeadm", "join", "--config '/etc/confi.yaml' --ignore-preflight-errors 'Service-Docker'"]' returned 1 instead of one of [0]
Error: /Stage[main]/Main/Exec[kubeadm join]/returns: change from 'notrun' to ['0'] failed: '["kubeadm", "join", "--config '/etc/confi.yaml' --ignore-preflight-errors 'Service-Docker'"]' returned 1 instead of one of [0]

## Expected Behavior
Should be like this
Debug: Exec[kubeadm join](provider=posix): Executing 'kubeadm join --config '/etc/kubernetes/config.yaml' --ignore-preflight-errors 'Service-Docker''
Debug: Executing: 'kubeadm join --config '/etc/kubernetes/config.yaml' --ignore-preflight-errors 'Service-Docker''

## Steps to Reproduce
install or join new nodes

## Environment
puppet-agent 6.28.0
puppet-server 6.20.0
Ubuntu focal

## Additional Context
Someone tested this commit e7addfbeaf3ff75496e31c2f44f22355f75dc6a3 ?
Why tests have been broken for more than two months ?
Broken build docker image  tooling.
I'm trying to fix it all in my PR https://github.com/puppetlabs/puppetlabs-kubernetes/pull/583

**Comment by BaronMsk** [comment]
I found that this commit broke a lot more than it seems, I suggest to git revert

**Comment by ArsenyBelorukov** [comment]
in addition, the wrong regexp was added in this commit https://github.com/puppetlabs/puppetlabs-kubernetes/commit/e7addfbeaf3ff75496e31c2f44f22355f75dc6a3#r86807409

**Comment by BaronMsk** [comment]
You can rollback this commit ?

**Comment by ArsenyBelorukov** [comment]
> You can rollback this commit ?

No

**Comment by BaronMsk** [comment]
Why, this makes the module not working ?

**Comment by ArsenyBelorukov** [comment]
Maybe @david22swan could help

**Comment by BaronMsk** [comment]
@GSPatton help please

**Comment by pmcmaw** [comment]
Hey @BaronMsk 
Apologies, I can see something has went wrong. We have taken note and this is something we are working on to resolve. 
In order to avoid issues like this arising in the future, I advise you to pull released versions of this module from the Forge. Our main branch may not always be as stable as we would like it to be. When a module is released it is our way of saying, this module has been thoroughly tested and it stable for end users to consume.

**Comment by BaronMsk** [comment]
@pmcmaw tnx

**Comment by LukasAud** [comment]
Hi @BaronMsk, we have been investigating this issue. Upon inspection of our original codebase hardening [PR](https://github.com/puppetlabs/puppetlabs-kubernetes/pull/575), we have been able to spot a few issue with our previous implementation and implement appropriate [fixes](https://github.com/puppetlabs/puppetlabs-kubernetes/pull/590) recently.

However, we are currently having issues investigating as we cannot replicate on our end.
The fix that was mentioned previously may be a potential fix for your issue but we are unable to confirm. Could you provide us with a more detailed 'Steps to reproduce' section in your issue report? Alternatively, would it be possible to test out the change?

**Comment by BaronMsk** [comment]
@LukasAud I can test your branch on my servers

**Comment by LukasAud** [comment]
Hi @BaronMsk, thanks for offering to test this. The current latest build we have in here in GitHub already has the patch I mentioned earlier. If you could update to our latest push and let me know if there is any visible impact in comparison to the failing version, that would be great.

**Comment by BaronMsk** [comment]
not work 
```Error: /Stage[main]/Kubernetes::Cluster_roles/Kubernetes::Kubeadm_join[k8s-test-worker-dtln-3]/Exec[kubeadm join]: Could not evaluate: Could not find command 'kubectl get nodes | grep k8s-test-worker-dtln-3'```

**Comment by LukasAud** [comment]
Thats interesting. I think I know what might the failure point for both of these errors. Both seem to be caused by my modification of the 'unless' commands. It seems wrapping unbreakable commands in arrays (and later, "fixed" by nesting those arrays) might have been counterproductive. 

It seems I will have to revert some of my updates. I still cannot be sure if this will help with the original issue but I will keep this thread updated as I work on the module.

**Comment by BaronMsk** [comment]
@LukasAud Yes, [commit](https://github.com/puppetlabs/puppetlabs-kubernetes/commit/e7addfbeaf3ff75496e31c2f44f22355f75dc6a3) revert will help 100%. I already checked it!

**Comment by LukasAud** [comment]
Unfortunately, this commit is important to our modules health and we cannot fully revert it. Our ideal scenario is to revert only whatever code is breaking the normal module behaviour and leave the rest. I will discuss this issue with a senior engineer in our team and see whats the best approach to fixing this.

**Comment by BaronMsk** [comment]
@LukasAud What problem did you want to solve with this commit?

**Comment by LukasAud** [comment]
@BaronMsk This commit was part of a larger project that was set to address concerns about code vulnerabilities in some of our modules. Unfortunately, some modules are giving us a harder time than others during this work. However, we consider this to be essential for the health of our modules and for establishing appropriate coding standards in the future.

**Comment by BaronMsk** [comment]
@LukasAud Where can I find these vulnerabilities ?

**Comment by LukasAud** [comment]
@BaronMsk This [article](https://dev.to/puppet/writing-secure-puppet-code-part-one-3dhm) we wrote should be able to explain the security issue better than me. More precisely, during our security update project, we have been targeting executable commands that contained interpolated user input. That involved examining 'command', 'onlyif' and 'unless' parameters and breaking them down where possible (this is not possible in certain scenarios, such as when a pipe | is present).

**Comment by LukasAud** [comment]
Hey @BaronMsk, we have merged this [bugfix](https://github.com/puppetlabs/puppetlabs-kubernetes/pull/592) which should address some of the latest issues. However, I cannot assure that this will fix the original problem. Right now, it looks like our team will have to take some time to dig deeper into the module and, most likely, do some maintenance before we can properly investigate the failure. 

Our forge build should still be a stable version of the module, so I would recommend sticking to that one for now. Sorry for the inconvenience.

**Comment by BaronMsk** [comment]
@LukasAud [it doesn't seem to work](https://github.com/puppetlabs/puppetlabs-kubernetes/blob/57eed2800a9100b60ac1eb8b07921305f4404ed9/manifests/packages.pp#L64) there can only be a string, otherwise, the expression is treated as different commands

**Comment by deric** [comment]
@BaronMsk I think #598 and #599 should fix most (if not all) of the problems mentioned above.

**Comment by chelnak** [comment]
+1 to the above - clusters seem to be spinning up from the HEAD of main.

**Comment by jordanbreen28** [comment]
@BaronMsk Thanks for raising this one, seems this issue has been resolved.
Are you able to test this out?

**Comment by BaronMsk** [comment]
@jordanbreen28 hi, i can check it in a few days.

**Comment by jordanbreen28** [comment]
Nice one @BaronMsk - if you could then update the thread here that'd be great!
